src/reader/arpa/mod.rs

```diff
@@ -0,0 +1,240 @@
+use itertools::Itertools;
+use std::{
+    io::{BufRead, BufReader, Lines},
+    num::NonZeroUsize,
+};
+
+use crate::headers::Counts;
+
+use super::{NGram, ProbBackoff, ProbBackoffNgram, ProbNgram};
+#[cfg(test)]
+mod test;
+
+#[derive(thiserror::Error, Debug)]
+pub enum ArpaReadError {
+    #[error("The /data/ header is missing")]
+    DataHeaderMissing,
+    #[error("Order key could not be parsed")]
+    NgramOrder,
+    #[error("NGram counts are missing in the \\data\\ section")]
+    NgramCountsMissing,
+    #[error("A NGram section with backoff is malformed.")]
+    BackOffSectionError,
+    #[error("A NGram section is missing its header.")]
+    NGramSectionHeaderMissing,
+    #[error("A NGram section mismatch. Got: {0}; Expected: {1}")]
+    NGramSectionHeaderMismatch(String, String),
+    #[error("actual NGram count does not match the header description.")]
+    NgramCountsMismatch,
+    #[error("A boundary between sections is missing. An empty line is expected")]
+    SectionBoundaryMissing,
+    #[error("The no-backoff section is malformed.")]
+    NoBackoffSectionError,
+}
+
+const ARPA_DATA_HEADER: &str = "\\data\\";
+const ARPA_NGRAM_KEY: &str = "ngram ";
+const ARPA_NGRAM_SECION_HEADERS: [&str; 7] = [
+    "\\1-grams:",
+    "\\2-grams:",
+    "\\3-grams:",
+    "\\4-grams:",
+    "\\5-grams:",
+    "\\6-grams:",
+    "\\7-grams:",
+];
+
+impl ProbNgram {
+    fn try_from_arpa_line(line: &str) -> Result<Self, ArpaReadError> {
+        let mut pieces = line.split_ascii_whitespace();
+        let log_prob = if let Some(Ok(log_prob)) = pieces.next().map(str::parse::<f32>) {
+            log_prob
+        } else {
+            return Err(ArpaReadError::NoBackoffSectionError);
+        };
+
+        let ngram = pieces.join(" ");
+
+        Ok(Self {
+            ngram: NGram(ngram),
+            prob: log_prob,
+        })
+    }
+}
+
+impl ProbBackoffNgram {
+    fn try_from_arpa_line(line: &str) -> Result<Self, ArpaReadError> {
+        let mut pieces = line.split_ascii_whitespace();
+        let log_prob = if let Some(Ok(log_prob)) = pieces.next().map(str::parse::<f32>) {
+            log_prob
+        } else {
+            return Err(ArpaReadError::BackOffSectionError);
+        };
+        let mut pieces = pieces.rev();
+        let backoff = if let Some(Ok(backoff)) = pieces.next().map(str::parse::<f32>) {
+            backoff
+        } else {
+            return Err(ArpaReadError::BackOffSectionError);
+        };
+
+        let ngram = pieces.rev().join(" ");
+
+        Ok(Self {
+            ngram: NGram(ngram),
+            prob_backoff: ProbBackoff { log_prob, backoff },
+        })
+    }
+}
+
+fn matches_ngram_section_header(line: &str, order: NonZeroUsize) -> Result<(), ArpaReadError> {
+    if (order.get() <= 7 && ARPA_NGRAM_SECION_HEADERS[order.get() - 1] != line)
+        || (order.get() > 7 && format!("\\{}-grams", order) != line)
```

Comment:
I don't think `format!()` is evaluated lazily here

Reply:
unnecessary optimization anyways, replaced everything with `format!`